app/modules/enduser_setup.c

```diff
@@ -625,6 +626,228 @@ static void do_station_cfg (task_param_t param, uint8_t prio)
+    if (data[pos] == '?' && (pos + 1 < data_len)) {
+      fields[counter] = &data[pos + 1];
+      break;
+    }
+  }
+
+  data[data_len] = 0; // this must be true by definition
+
+  for (pos = 0; pos < data_len && counter < MAX_HTTP_FIELDS && state != STATE_FIN; pos++) {
+    //std::cout << "val" << pos << ":" << data[pos] << std::endl;
+    if (data[pos] == '&' || data[pos] == ';') { // w3c states & or ; separate params
+      if (lastPos != -1) {
+        data[pos] = 0; // terminate the string
+        counter++;
+
+        if (counter < MAX_HTTP_FIELDS) {
+          fields[counter] = NULL;
+          values[counter] = NULL;
+        }
+        state = STATE_FIELD;
+      } else { // invalid state, started with &, skip it
+
+      }
+    } else if (state == STATE_FIELD) {
+      if (data[pos] == '=') {
+        state = STATE_DATA;
+        data[pos] = 0;
+      } else {
+        if (fields[counter] == NULL) {
+          fields[counter] = &data[pos];
+        }
+      }
+    } else if (state == STATE_DATA) {
+      // ensure we deal with "field=" (i.e. no value)
+      if (values[counter] == NULL) {
+        values[counter] = &data[pos];
+      }
+
+      if (data[pos] == ' ') { // at the end - this is invalid query param, space is %20
+        counter ++;
+        data[pos] = 0;
+        state = STATE_FIN;
+      }
+    }
+
+    lastPos = pos;
+  }
+
+  // we always fall out with one less than the actual number of fields
+  if (state == STATE_DATA) {
+    counter ++;
+  }
+
+  if (counter >= MAX_HTTP_FIELDS) {
+    counter = MAX_HTTP_FIELDS-1;
+  }
+
+  return counter;
+}
+
+typedef struct enduser_Validation_Result {
+  int status;
+  const char* body;
+  const char* content_type;
+} enduser_Validation_Result;
+
+static void enduser_extract_validation_result(lua_State *L, enduser_Validation_Result *val) {
+  int idx = lua_gettop( L );
+
+  lua_getfield( L, idx, "status" );
+  lua_getfield( L, idx, "content_type");
+  lua_getfield( L, idx, "body" );
+
+  val->status = luaL_optint( L, ++idx, 200 );
+  val->content_type =  luaL_optlstring(L, ++idx, NULL, 0);
+  val->body =  luaL_optlstring(L, ++idx, NULL, 0);
+
+  if (val->content_type) {
+    val->content_type = c_strdup(val->content_type);
+  }
+
+  if (val->body) {
+    val->body = c_strdup(val->body);
+  }
+
+  lua_pop( L, 3 );
+}
+
+static void enduser_cleanup_validation_result(enduser_Validation_Result* val) {
+  if (val->body != NULL)
+  {
+    os_free((void *)val->body);
+  }
+  if (val->content_type != NULL)
+  {
+    os_free((void *)val->content_type);
+  }
+}
+
+static int enduser_find_wifi_pwd_loc(int param_count, char *fields_str[],
+               char *values_str[], enduser_Validation_Result *validation,
+               int *wifi_ssid_loc, int *wifi_pwd_loc) {
+  int p_file = 0;
+  lua_State *L = NULL;
+
+  if (param_count > 2)
+  {
+    ENDUSER_SETUP_DEBUG("enduser: opening enduser.json for write");
+
+    // setup the file output
+    p_file = vfs_open("enduser.json", "w");
```

Comment:
You're right, it may not be 100% logical but "short" is sometimes the opposite of "meaningful" 😉 

My point was that `enduser.json` is IMHO too short or too generic. You can't deduce the content of the file from its name.